test/spec_session_cookie.rb

```diff
@@ -262,69 +283,98 @@ def decode(str); @calls << :decode; str; end
     response = response_for(:app => app, :cookie => response)
     response.body.must_equal '{"counter"=>3}'
 
-    app = [incrementor, { :secret => "other" }]
+    app = [incrementor, { :secret => random_cipher_secret }]
 
     response = response_for(:app => app, :cookie => response)
     response.body.must_equal '{"counter"=>1}'
   end
 
   it "loads from a cookie with accept-only integrity hash for graceful key rotation" do
-    response = response_for(:app => [incrementor, { :secret => "test" }])
+    response = response_for(:app => [incrementor, { :secret => @secret }])
+
+    new_secret = random_cipher_secret
 
-    app = [incrementor, { :secret => "test2", :old_secret => "test" }]
+    app = [incrementor, { :secret => new_secret, :old_secret => @secret }]
     response = response_for(:app => app, :cookie => response)
     response.body.must_equal '{"counter"=>2}'
 
-    app = [incrementor, { :secret => "test3", :old_secret => "test2" }]
+    newer_secret = random_cipher_secret
+
+    app = [incrementor, { :secret => newer_secret, :old_secret => new_secret }]
     response = response_for(:app => app, :cookie => response)
     response.body.must_equal '{"counter"=>3}'
   end
 
+  it 'loads from a legacy hmac cookie' do
+    legacy_session = Rack::Session::Cookie::Base64::Marshal.new.encode({ 'counter' => 1, 'session_id' => 'abcdef' })
+    legacy_secret  = 'test legacy secret'
+    legacy_digest  = OpenSSL::HMAC.hexdigest(OpenSSL::Digest::SHA1.new, legacy_secret, legacy_session)
+
+    legacy_cookie = "rack.session=#{legacy_session}--#{legacy_digest}; path=/; HttpOnly"
+
+    app = [incrementor, { :secret => @secret, :legacy_hmac_secret => legacy_secret }]
+    response = response_for(:app => app, :cookie => legacy_cookie)
+    response.body.must_equal '{"counter"=>2}'
+  end
+
   it "ignores tampered with session cookies" do
-    app = [incrementor, { :secret => "test" }]
+    app = [incrementor, { :secret => @secret }]
     response = response_for(:app => app)
     response.body.must_equal '{"counter"=>1}'
 
     response = response_for(:app => app, :cookie => response)
     response.body.must_equal '{"counter"=>2}'
 
-    _, digest = response["Set-Cookie"].split("--")
-    tampered_with_cookie = "hackerman-was-here" + "--" + digest
+    ctxt, iv, auth_tag = response["Set-Cookie"].split("--", 3)
+    tampered_with_cookie = [ctxt, iv, auth_tag.reverse].join('--')
 
     response = response_for(:app => app, :cookie => tampered_with_cookie)
     response.body.must_equal '{"counter"=>1}'
   end
 
+  it 'ignores tampered with legacy hmac cookie' do
+    legacy_session = Rack::Session::Cookie::Base64::Marshal.new.encode({ 'counter' => 1, 'session_id' => 'abcdef' })
+    legacy_secret  = 'test legacy secret'
+    legacy_digest  = OpenSSL::HMAC.hexdigest(OpenSSL::Digest::SHA1.new, legacy_secret, legacy_session).reverse
+
+    legacy_cookie = "rack.session=#{legacy_session}--#{legacy_digest}; path=/; HttpOnly"
```

Comment:
Q: Below, the Rack::Utils.escape was used on the legacy_session - not here. Does this matter?

Reply:
We actually don't need `escape` when an encoder with strict base64 encoding is used.